openpower/patches/vesnin-patches/petitboot/petitboot-1001-discover-platform-powerpc-fix-bug-with-ipmi-boot-mai.patch

```diff
@@ -0,0 +1,280 @@
+@@ -437,10 +460,10 @@ static int get_ipmi_bootdev_ipmi(struct platform_powerpc *platform,
+ }
+ 
+ static int get_ipmi_boot_mailbox_block(struct platform_powerpc *platform,
+-		char *buf, uint8_t block)
++		mbox_t *mailbox, uint8_t block)
+ {
+-	size_t blocksize = 16;
+-	uint8_t resp[3 + 16];
++	size_t blocksize;
++	ipmi_mbox_response_t ipmi_mbox_resp;
+ 	uint16_t resp_len;
+ 	char *debug_buf;
+ 	int rc;
+@@ -450,59 +473,68 @@ static int get_ipmi_boot_mailbox_block(struct platform_powerpc *platform,
+ 		0x00,  /* no block selector */
+ 	};
+ 
+-	resp_len = sizeof(resp);
++	resp_len = sizeof(ipmi_mbox_response_t);
+ 	rc = ipmi_transaction(platform->ipmi, IPMI_NETFN_CHASSIS,
+ 			IPMI_CMD_CHASSIS_GET_SYSTEM_BOOT_OPTIONS,
+ 			req, sizeof(req),
+-			resp, &resp_len,
++			(uint8_t*)&ipmi_mbox_resp, &resp_len,
+ 			ipmi_timeout);
+ 	if (rc) {
+ 		pb_log("platform: error reading IPMI boot options\n");
+ 		return -1;
+ 	}
+ 
+-	if (resp_len < sizeof(resp)) {
+-		if (resp_len < 3) {
+-			pb_log("platform: unexpected length (%d) in "
+-					"boot options mailbox response\n",
+-					resp_len);
+-			return -1;
+-		}
+-
+-		if (resp_len == 4) {
+-			pb_debug_fn("block %hu empty\n", block);
+-			return 0;
+-		}
++	if (resp_len > sizeof(ipmi_mbox_response_t)) {
++		pb_debug("platform: invalid mailbox response size!\n");
++		return -1;
++	}
+ 
+-		blocksize = sizeof(resp) - 3;
+-		pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
+-				block, blocksize);
++	if (resp_len < 3) {
++		pb_log("platform: unexpected length (%d) in "
++				"boot options mailbox response\n",
++				resp_len);
++		return -1;
+ 	}
+ 
+-	debug_buf = format_buffer(platform, resp, resp_len);
++	blocksize = resp_len - (sizeof(ipmi_mbox_response_t) - sizeof(mbox_t));
++	pb_debug_fn("Mailbox block %hu returns only %zu bytes in block\n",
++			block, blocksize);
++
++	debug_buf = format_buffer(platform, (uint8_t*)&ipmi_mbox_resp, resp_len);
+ 	pb_debug_fn("IPMI bootdev mailbox block %hu:\n%s\n", block, debug_buf);
+ 	talloc_free(debug_buf);
+ 
+-	if (resp[0] != 0) {
++	if (ipmi_mbox_resp.cc != 0) {
+ 		pb_log("platform: non-zero completion code %d from IPMI req\n",
+-				resp[0]);
++				ipmi_mbox_resp.cc);
+ 		return -1;
+ 	}
+ 
+ 	/* check for correct parameter version */
+-	if ((resp[1] & 0xf) != 0x1) {
++	if ((ipmi_mbox_resp.param_version & 0xf) != 0x1) {
+ 		pb_log("platform: unexpected version (0x%x) in "
+-				"boot mailbox response\n", resp[0]);
++				"boot mailbox response\n", ipmi_mbox_resp.param_version);
+ 		return -1;
+ 	}
+ 
+ 	/* check for valid paramters */
+-	if (resp[2] & 0x80) {
++	if (ipmi_mbox_resp.param_valid & 0x80) {
+ 		pb_debug("platform: boot mailbox parameters are invalid/locked\n");
+ 		return -1;
+ 	}
+ 
+-	memcpy(buf, &resp[3], blocksize);
++	if (block != ipmi_mbox_resp.block_selector) {
++		pb_debug("platform: returned boot mailbox block doesn't match "
++				  "requested\n");
++		return -1;
++	}
++
++	if (blocksize == 0) {
```

Comment:
I prefer `!blocksize` notation for zero checking. But this is of course ok too.

Reply:
Done